lexer.ebnf

```diff
@@ -0,0 +1,122 @@
+(* mash token literal sepcification *)
```

Comment:
Not sure why, but I feel like this grammar is not correct and quite ambiguous. I'm not sure if I asked this before but if you haven't already, can you create a proper ebnf document (for both command and not command) route and run it through some ebnf to parser generators?

Reply:
The two `.ebnf` files together do create a proper ebnf specification. I have never exactly used parser generators, so it will be better if you could do it.

Reply:
I think they are quite straight forward (at least Bison or Yacc is). If you want I can help you out but I think it would be best for you to get familiar with them because Bison is very good at making sure that their are no ambiguities in your bnf.

The only problem that I see is that you have EBNF not BNF and bison requires BNF. But there are other compiler-compilers that are able to generate a AST parser from EBNF - I don't have experience with them however so won't be able to help with that.

Reply:
Maybe [this](https://github.com/blynn/nex) will come in handy for go code (although I haven't used it myself)

Reply:
Converting it to bnf should be easy, it will just be a tedious job involving a lot of recursion.

Reply:
@tjgurwara99 http://lampwww.epfl.ch/teaching/archive/compilation-ssc/2000/part4/parsing/node3.html

Reply:
What about it? @raklaptudirm

Reply:
A guide of how to convert `ebnf` to `bnf`

Reply:
Lol, I can see that but whats that got to do with what I said haha.

Anyways, if you want me to do it, I'm more that happy to but I still think it's better for you to do it yourself (making a parser from scratch is not always viable). So learning one of the parser generators would be very useful, though I'm not sure if you're familiar with C/C++ parser generators so won't recommend any specific parser here. I personally use bison for log files parsing (thats the most likely usecase you will see for these generators).

If you just want to check whether your syntax is correct, I read about a tool a few months back [here](https://mdkrajnak.github.io/ebnftest/) this lets you see your parser in action. So if your grammar is correct it should work.

Reply:
Looks like it does not ignore whitespace XD. I am having to format test cases like this:
```
letn:=6;letf:=1;leti:= 2;fori<=n{letf*=i;leti+=1;};echoFactorialof[n]=[f];end of file
```
Any ways:
```
Test Output

The test input is valid.
```

Reply:
Just add an extra token in the middle like this

```
WhiteSpace   ::= ' ' | '\t' | '\n'
```
and add this whitespace in the middle of the all the other tokens combinations - for example, `LetStatement   ::= LetKeyword WhiteSpace Assign Equal  WhiteSpace Integer`

Also, I think I mentioned this before but don't remember whether you did it or not - you should write a baseline program in this new language to make sure that this will indeed work and put that in this ebnf checker.

Reply:
I will try to learn bison, but in the meantime, if you can, convert the grammar to bnf and try to use bison to parse it.

Reply:
@tjgurwara99

Reply:
Will take me some time - but if it worked on the link I mentioned then it should be fine. But you didn't answer my other part of the comment, have you create an actual program that you think should work in the language so that we can consider it as a baseline

Reply:
I tested the parser using the factorial program:
```
let n := 6
let f := 1
let i := 2

for i <= n {
  let f *= i
  let i += 1
}

echo Factorial of [n] = [f]
```

Reply:
You can merge this - I tested it but I need to know a few more things before I could get it to work, I'm rusty on Bison so its expected. I think online parser check is enough